src/cascadia/TerminalApp/CascadiaSettings.cpp

```diff
@@ -719,6 +721,21 @@ std::string CascadiaSettings::_ApplyFirstRunChangesToSettingsTemplate(std::strin
     return finalSettings;
 }
 
+// Method Description:
+// - Expands any commands with `iterateOn` set. If we successfully generated
+//   expanded commands for them, then we'll remove the original command, and add
+//   all the newly generated commands.
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void CascadiaSettings::_ExpandCommands()
+{
+    winrt::TerminalApp::implementation::Command::ExpandCommands(_globals.GetCommands(),
+                                                                _profiles,
+                                                                _warnings);
+}

```

Comment:
I'm worried about the impact this'll have on the settings model. Should we keep command expansion out of Settings and up in the App/Logic layer?

Reply:
Well the settings model will have a full understanding of all the profiles, schemes, themes - all the things that could result in command expansion, right? So it probably makes more sense to keep it _within_ the settings object, so that it can be pulled out easier?

Reply:
Sorta? Dynamic profiles are an interesting case of “mostly outside the settings model”, and that’s something we need a comprehensive settings model spec for. Having the application do expansion helps clean up responsibility a little bit.

Oh, also, what about expansions over the list of open tabs? We don’t want a settings reload for that...

Reply:
>  what about expansions over the list of open tabs

I hadn't even considered that yet as a possibility for command expansion. Do we have a hypothetical command for that?

Reply:
Maybe “switch to first instance of” or “dock split pane from open tabs”

Reply:
Okay
1. ew
2. I suppose `iterateOn: tabs`/`iterateOn:panes` could be special and only expanded at runtime, given a list of tabs/panes. It would need to be expanded when the command palette is opened. The settings model wouldn't be able to expand that object on it's own, since it wouldn't have a list of tabs. 

I don't know, that's just brainstorming

Reply:
Okay let me explore how horrifying _expanding iterable commands when the palette is opened_ is. For things like profiles, this seems like it'll be strictly worse, but that's the only way to expand them for things like tabs.

Maybe we can cache the results of the expansion, and only do the re-expansion when the number of tabs changes?

@DHowett/@carlos-zamora am I right in thinking you two are blocking till this issue is resolved?

Reply:
Hokay, so I _think_ I've got it re-expanding the list of commands each time the list of tabs changes, and I don't think it's noticable. 

I hate it, because it involves having `TerminalPage` _make a copy_ of the commands from `CascadiaSettings`, then expand that copy, every time the list of tabs changes. But on the bright side, `CascadiaSettings` won't need to know anything about the list of tabs when the expansion happens. `CascadiaSettings` would only keep the original blob of commands around, and they'd _need_ to be expanded before the palette could use them. This seems not horrible.

The copying is going to be more annoying when the map of commands isn't just an `unordered_map`. We'd need to be able to clone the original `winrt::...Map` into a new one, then expand the commands using only methods that are exposed by the _projection_. Assuming that we want to have `ExpandCommands` eventually take a list of the current tabs, this means that the projected `ExpandCommands` would either need to:
* be implemented in `Microsoft::Terminal::Settings::Command`, and accept a `TerminalApp::Tabs`. This would be a layering violation. 
  - We could work around this by having our tabs expose something dumb like `Terminal::Settings::ITabInformation` and have `ExpandCommands` accept that instead.
    - this is the solution I'd be _least_ opposed to
* be implemented in `TerminalApp`, but then we wouldn't be able to inspect the implementation details of `Command`, which means we wouldn't be able to easily access the original json used for the command (which we need to be able to expand the command).
  - we could do something extra dumb like exposing the original json as a projected method on `Command`, but that's definitely silly

So I've pushed just this change to [`dev/migrie/f/6856-let-terminalpage-expandcommands`](https://github.com/microsoft/terminal/compare/dev/migrie/f/command-palette-iterateOn...dev/migrie/f/6856-let-terminalpage-expandcommands?expand=1). This doesn't have the madness above included, obviously. We can keep it if people agree it's good

Reply:
I think that's fine. Let's go the first route.